Generate AppImage

andyleejordan opened the pull request:
Replaces #2027 as it had merge conflicts due to macOS -> OS X rename. Rebased to fix failing Pester test run.

@probonopd and @TravisEz13 please review that I fixed conflicts correctly.

Push build is [here](https://travis-ci.org/andschwa/PowerShell/builds/155445453).

probonopd committed:
Add AppImage tooling

* Use icon from local repository
  https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960616

* Use the deb that has been generated in this build

* Copyright and license
  https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960505

* MIT License for appimage.sh
  https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960505

* Full text of the MIT License is in license_thirdparty_proprietary.txt

* Clarify license and clean up unused code
  https://github.com/PowerShell/PowerShell/pull/2027#discussion_r76104601

* Mark appimage.sh as executable

probonopd committed:
Add AppImage documentation

msftclas commented:
Hi __@andschwa__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Andy Schwartzmeyer). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

andschwa commented:
Push build succeeded. Pester tests ran and AppImage was created successfully.

probonopd commented:
Thanks, looks good to me.

andschwa commented:
@TravisEz13 can you try out the AppImage?

andschwa commented:
@TravisEz13 ah, of course. Here's an upstream [build](https://travis-ci.org/PowerShell/PowerShell/builds/155773444) which should upload the artifacts when its complete.

probonopd commented:
Upload succeeded, download_url: https://s3.amazonaws.com/powershell-artifacts/PowerShell/PowerShell/258/258.1/powershell-6.0.0-alpha.9-x86_64.AppImage

Will test now.

probonopd commented:
Tested on Ubuntu 16.04, runs.

andschwa commented:
LGTM. @TravisEz13?

andschwa commented:
@probonopd I have a hold that came internally, we can't merge this yet.

@KrisBash can you please explain?

andschwa commented:
@KrisBash nevermind, I spoke with @probonopd.

lzybkr commented on tools/appimage.sh:
```diff
@@ -0,0 +1,141 @@
+
+cd ./$APP/
+
+wget -q https://github.com/probonopd/AppImages/raw/master/functions.sh -O ./functions.sh
```
We shouldn't be fetching scripts from potentially untrusted sources as part of building our package - we should have copies of everything we need in our repo.

probonopd replied:
functions.sh is part of the [probonopd/AppImages](https://github.com/probonopd/AppImages) project which is under the MIT License so you are free to mirror it but it is neither part of the PowerShell project nor subject to the Microsoft CLA. The binaries it downloads in the generate_appimage function are built from the sources in the the [probonopd/AppImageKit](https://github.com/probonopd/AppImageKit) project which is also under the MIT License but is also neither part of the PowerShell project nor subject to the Microsoft CLA. You are free to rebuild them from your own mirror, too, if you like.

andschwa commented:
Hm, I can't assign @KrisBash to the issue.

@lzybkr should we add him as a collaborator?

zaxebo1 commented:
I am using powershell AppImage on linux , available at https://bintray.com/probono/AppImages/PowerShell .  It helps in single portable binary executable file getting executed on various linux distributions. 
**I will humbly request and very strongly recommend that each powershell release by the powershell development team, should also officially happen in AppImage format , and should be available at https://github.com/PowerShell/PowerShell/releases**

andschwa commented:
Hello, and thank you for your input @zaxebo1. I am following up internally on why this is still blocked; we should have a better answer here soon. Thanks again :smile:

SteveL-MSFT commented:
Working on getting resolution on this, won't happen til next week earliest

andschwa commented:
Thanks for following up @SteveL-MSFT 😄

SteveL-MSFT commented:
@andschwa this has been approved, but we must include some attribution, I'll send them to you via email to include in the image

zaxebo1 commented:
the above are **trivial conflicts** - that is, conflicts only in markdown files and a shell script, which can be quickly resolved. Request you to kindly take up  on early/priority - to resolve these conflicting files

probonopd commented:
@zaxebo1 I think I don't have the permissions, only @andschwa can do this since he opened this PR based on a previous one from me.

andschwa commented:
@probonopd is right, since I can fix this up next week.

vors commented:
@andschwa ping

andschwa commented:
@vors is there a !remindme bot for GitHub? If not, there should be. Adding this to my task list...

vors commented:
I will play its role for now :)

andyleejordan committed:
Add AppImage to CI

andschwa commented:
@vors this is ready now as far as I know.

andschwa commented:
Steve can you send the attributions requirements again?

andyleejordan committed:
Add Third Party Notices for AppImage

This license specifically covers the additional libraries that
are bundled inside the AppImage archive.

andschwa commented on README.md:
```diff
@@ -39,13 +40,15 @@ You can download and install a PowerShell package for any of the following platf
 [rl-ubuntu16]: https://github.com/PowerShell/PowerShell/releases/download/v6.0.0-alpha.15/powershell_6.0.0-alpha.15-1ubuntu1.16.04.1_amd64.deb
 [rl-ubuntu14]: https://github.com/PowerShell/PowerShell/releases/download/v6.0.0-alpha.15/powershell_6.0.0-alpha.15-1ubuntu1.14.04.1_amd64.deb
 [rl-centos]: https://github.com/PowerShell/PowerShell/releases/download/v6.0.0-alpha.15/powershell-6.0.0_alpha.15-1.el7.centos.x86_64.rpm
+[rl-ai]: TODO: Post link when availabl
```
Oops a typo.

andschwa commented:
- [x] Rebased and merge conflicts resolved
- [x] Pinned external code download to specific commit hash
- [x] Included additional third party licenses from legal
- [x] Successful build of AppImage binary
- [ ] Anything else?

@PowerShell/powershell-maintainers you'll want to include the AppImage binary in the next release on GitHub, and fix the link in the readme.

andschwa commented:
@SteveL-MSFT handing over to you for final review and merge 😄

joeyaiello commented:
> Anything else?

@andschwa just *another* rebase ;)

andschwa commented:
@joeyaiello your turn!

probonopd commented:
Never ending story ;-) @andschwa

SteveL-MSFT reviewed (approved):
LGTM

probonopd commented:
Thanks for merging. Minor cosmetic glitch, in `README.md` it now says `[.AppImage][rl-ai]` which doesn't look right.

SteveL-MSFT commented:
@raghushantha do you need to update your publishing pipeline to generate the appimage?  once we have an official appimage, can you update the link to point to it?

andschwa commented:
Woo hoo!

probonopd commented:
Let's celebrate @andschwa :1st_place_medal:

raghushantha commented:
@SteveL-MSFT I have integrated AppImage generation as part of the Release pipeline.
Got a couple of concerns which I will followup offline
